### ISSUE TITLE
Bug 412391: Add support for hidden mappedsuperclass attributes

### ISSUE THREAD

**Issue by dazey3** [eclipse-ee4j/eclipselink#1327]
Linking to [BZ #412391](https://bugs.eclipse.org/bugs/show_bug.cgi?id=412391)

Environment:
Eclipselink 2.3.2
Static weaving turned on

```java
@MappedSuperclass
public abstract class A {

    @Column(name = "CREATED_AT")
    @Temporal(TemporalType.TIMESTAMP)
    @NotNull
    private Date createdAt = new Date();

(...) getters, setters...
}

public class B extends A {
    @Column(name = "CREATED_AT")
    @Temporal(TemporalType.TIMESTAMP)
    private Date createdAt;

(...) getters, setters...
}

public class C extends A {

// note: no createdAt attribute defined here!
}
```

Now the following lines sometimes fail and sometimes don't:
```java
C c = new C();
entityManager.persist(c)
```

with "com.mysql.jdbc.exceptions.jdbc4.MySQLIntegrityConstraintViolationException: Column 'CREATED_AT' cannot be null" even though {{@NotNull}} annotation was used and new Date() was applied.

Please note, I am NOT persisting B here at all but the sole existence of B interferes with persisting of C!

== Root cause ==

Now, after having debugged eclipselink we found out that ((PersistenceObject) c)._persistence_get("createdAt") return null even if one uses interned String.
That obviously makes eclipselink believe that createdAt is null even if it is not.

== Solution ? ==
Now, we should either fix this in {{ClassWeaver.addPersistenceGetSet()}} method. 

An alternative would be to make such a setting (overriding class attribute) forbidden and failing on startup time.
It took us several days to figure out what actually goes wrong here...

**Comment by dazey3** [comment]
@TomWare
> You should be able to work around an issue in _persistence_get()/_persistence_set() by setting persistence unit property:
> eclipselink.weaving.internal=false

I confirmed this does work around the issue by not using the weaved get/set methods

**Comment by dazey3** [comment]
It's been almost a decade since this issue was originally reported, but it appears that it may still be an issue. I believe I am hitting a variation of this same issue, but with dynamic weaving. The result is that child values, inheriting from the superclass, are null after an em.find()

Entities:
```java
    @MappedSuperclass
    public abstract class AbstractBE {
        private String parentOnlyAttribute;
        private Short hiddenAttribute;
    }

    @Entity
    public class EntityA extends AbstractBE {
        @Id private long id;
        private Short hiddenAttribute;
    }
```

Setup:
```SQL
    INSERT INTO ENTITYA (ID, HIDDENATTRIBUTE, PARENTONLYATTRIBUTE) VALUES (123456, 1, 'ParentValue')
```
```
    -Declipselink.weaving.output.path=... -javaagent:.../eclipselink.jar
```
    persistence.xml:
```xml
        <property name="eclipselink.weaving" value="true" />
```

Test:
```java
    EntityManager em = emf.createEntityManager();
    EntityA e = em.find(EntityA.class, 123456l);
```

Result: e.getHiddenAttribute() is NULL.

The issue is in the weaved code for AbstractBE. In the weaved code, EclipseLink generates the method `_persistence_set(String paramString, Object paramObject)` that is responsible for setting the attribute values on the object. Except, the "hiddenAttribute" attribute is not there. This means later sets on Classes of the parent type will NOOP, leaving them null!

Tracing it:

(org.eclipse.persistence.internal.jpa.weaving.TransformerFactory.storeAttributeMappings)

EclipseLink scans the metadata for EntityA and determines EntityA has 3 mappings: "id", "hiddenAttribute", & "parentOnlyAttribute". The mapping "parentOnlyAttribute" isn't found in EntityA, so it's added to "unMappedAttributes" for further processing in the superclass. The mapping "hiddenAttribute" though is found on EntityA, so AbstractBE never gets it too!

**TLDR;** EclipseLink determines ClassDetails@1[EntityA]{"id", "hiddenAttribute"} -> ClassDetails@2[AbstractBE]{"parentOnlyAttribute"}

--- Expanding the test ---

Now let's consider adding a new class similar to the original post

```java
    @Entity
    public class EntityB extends AbstractBE {
        @Id private long id;
    }
```
This class doesn't hide any attribute values, but em.find(EntityB.class, ...) will still result in `getHiddenAttribute()` returning NULL!

EclipseLink scans the metadata for EntityB and determines EntityB has 3 mappings: "id", "hiddenAttribute", & "parentOnlyAttribute". However, this time the mappings "parentOnlyAttribute" & "hiddenAttribute" aren't found in EntityB, so they're added to "unMappedAttributes" for further processing in the superclass. They both get added to AbstractBE ClassDetails

**TLDR;** EclipseLink determines 
ClassDetails@1[EntityA]{"id", "hiddenAttribute"} -> ClassDetails@2[AbstractBE]{"parentOnlyAttribute"}
ClassDetails@3[EntityB]{"id"} -> ClassDetails@4[AbstractBE]{"parentOnlyAttribute", "hiddenAttribute"}

Notice that EntityA ClassDetails is mapped an instance of ClassDetails for AbstractBE, but EntityB ClassDetails is mapped to a different instance of ClassDetails for AbstractBE! At weave time, EntityA's AbstractBE ClassDetails end up being used to write out.
